### ISSUE TITLE
[SCIONLab] Renaming and refactoring

### ISSUE THREAD

**Issue by mlegner** [netsec-ethz/scion-coord#97]
Some of the terminology used in SCIONLab got more and more confusing over time.
As discussed with @ercanucan and others, we should do some refactoring and rename databases, variables, etc.
I propose the following naming conventions [**NewName** (OldName): Description]:

- **AttachmentPoint (AP)** (SCIONLabAS/SCIONLabServer): SCION ASes to which new ASes can be attached automatically
- **SCIONLabAS** (SCIONLabVM): ASes connected to APs that are set up and administered by users
- **SCIONBox**: Machines that are automatically managed through scion-coord and connected to APs;
they can be APs themselves if the connection is good enough

Do you agree with these conventions or suggest different/additional ones?
Another question is what the relationship between SCIONLabASes and SCIONBoxes should be:
Is one a subset of the other or should they be separate concepts?

**Comment by philippmao** [comment]
I think the SCIONLabASes and SCIONBoxes are similar concepts. But I think the two should be treated separately just because they will be treated differently by the coordinator. The SCIONLabAS has only one connection to an AP. Whereas the SCIONBox can be connected to multiple AP. Also the SCIONLabAS and the SCIONBoxes interact quite differently with the coordinator. 
Please have a look at my models/scion_box.go where I implemented a first idea of the SCIONBox database:  https://github.com/philippmao/scion-coord/blob/master/models/scion_box.go

**Comment by ercanucan** [comment]
@mlegner Thanks for the suggestions. As for the connection between SCIONLabAS and SCIONBox, I think SCIONBox can be considered as a sub-type of SCIONLabAS. In my opinion, SCIONLabAS can be the "umbrella" term, and we can have sub-types of it, such as the following:
- **SCIONLab VM AS**: ASes running inside virtual machines.
- **SCIONLab Native AS** (or another word if native sounds misleading): ASes running inside dedicated PCs, setup via the pure-gen folder feature.
- **SCIONBox**: ASes running inside PCEngines boxes.

Futhermore, we can introduce a concept called **Role**. A SCIONLabAS can come in various roles. One of such roles can be **AttachmentPoint (AP)**.

What do you think?

**Comment by hausheer** [comment]
@ercanucan I like the concept of roles and the notion of a SCIONLab AS that can be instantiated in different ways. An alternative name for SCIONLab Native AS could be SCIONLab Dedicated AS (as suggested in your description)
@philippmao A SCIONLab VM AS may have multiple connections to APs in future as well. 
Let's document all these terms in the SCIONLab paper.

**Comment by mlegner** [comment]
Thank you for all the suggestions. I like the term `Dedicated AS` and the `role` attribute.

I guess the distinction between **Dedicated** and **VM** is not so big on the code level; there it would suffice to have a flag similar to the boolean flag `isDedicated` introduced in PR #73. Do you agree?

On the database level, we could then have one table for _all_ SCIONLab ASes (containing a `type` attribute VM/Dedicated/Box replacing the boolean flag) and a separate table keeping all additional information needed for SCIONBoxes.

I suggest doing this refactoring as soon as PRs #73 and #95 are merged such that all is ready for the SCIONBoxes. As Philipp is mostly responsible for those, he could also do the refactoring before. Any objections?

**Comment by FR4NK-W** [comment]
Yes, this sounds good. I also think that the main difference between **Dedicated** and **VM** will be the availability and performance one can expect.

**Comment by philippmao** [comment]
How about something like this ?
```
type SCIONLabAS struct {
         ID             uint64     
         UserEmail      string
         IA             *AS
         IP             string
         type           VM/DEDICATED/BOX
         AP             bool                   // is this AS an attachment point?
         Status         uint8
         MAC            string                 // only used for SCIONBoxes
         B-Routers      []BorderRouter         // keep track of neighbors and assigned ports
}
```
```
type BorderRouter struct {
        IID             uint64       // Interface-ID
        IA              *AS          // neighbor IA
        IP              string       // IP address of the neighbor border router
        RemotePort      string
        LocalPort       string
        Linktype        string       // PARENT, CHILD, CORE
        isVPN           bool
}
```
The SCIONLabServer will probably have a struct similar to the SCIONLabAS or we might even put the SCIONLabServer into the same struct.

**Comment by hausheer** [comment]
I think the main difference between Dedicated and VM is the Vagrant stuff that is needed for the VM additionally.
@philippmao By SCIONLabServer you mean SCIONLab Attachment Point?

**Comment by philippmao** [comment]
@hausheer
Yes, I was just wondering if we want to treat a scionlabserver attachement point differently than for example a scionbox attachement point.

**Comment by hausheer** [comment]
@philippmao As suggested by @ercanucan attachment point can be a role, that in my view any SCIONLab AS could have, primarily a Dedicated AS or a SCIONBox but potentially even a VM.

**Comment by mlegner** [comment]
This is partially fixed by PR #121.

**Comment by mlegner** [comment]
Fixed by PR #133.
